core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -3774,11 +3780,130 @@ protected void validateJoin(SqlJoin join, SqlValidatorScope scope) {
             RESOURCE.crossJoinDisallowsCondition());
       }
       break;
+    case LEFT_ASOF:
+    case ASOF: {
+      // In addition to the standard join checks, the ASOF join requires the
+      // ON conditions to be a conjunction of simple equalities from both relations.
+      SqlAsofJoin asof = (SqlAsofJoin) join;
+      SqlNode matchCondition = getMatchCondition(asof);
+      matchCondition = expand(matchCondition, joinScope);
+      join.setOperand(6, matchCondition);
+      validateWhereOrOn(joinScope, matchCondition, "MATCH_CONDITION");
+      ConjunctionOfEqualities conj = new ConjunctionOfEqualities();
```

Comment:
Would it be better to replace RelOptUtil.conjunctions() and match equal here.
Maybe the bool filed type affects like bool_column3 = (bool_column1 and bool_column2). WDYT

Reply:
Of course, this is a rare case

Reply:
"and' is not the same as "=" for booleans, since "and" returns "false" for "false = false".
But maybe I misunderstand your question.

Reply:
I'm guessing that the checks here refer to snowflake's specification,  i.e. "ON clause for ASOF JOIN must contain conjunctions of equality conditions only. Each side of an equality condition must only refer to either the left table or the right table. (S.STATE = P.STATE) OR (S.LOCATION = P.LOCATION) is invalid."
We expect that condition needs to satisfy 
`l.c1 = r.c1 and l.c2 = r.c2`
But if you just check if sqlcall is equal / and, may a case like the following would also pass, if the types of c1 and c2 would both be boolean
`(l.c1 and r.c1 ) and  (l.c2 and r.c2 )`

Reply:
Good point. I will add a test case to make sure this doesn't happen.

Reply:
Unfortunately I can't reuse RelOptUtil because that works on RexNode, and here I only have SqlNode objects.
I will improve the ConjunctionOfEqualities visitor.